docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+chance of leadership changes or node restarts before we reach it.
+
+If we wanted to support very long idle timeouts better, we could
+do so by adding something like a kvstore record of the access time
+that is updated at most every N minutes, and when a leader is elected
+it primes its access time from that if it is the same node that was
+also the previous leader.
+
+## What resources are we saving by hibernating?
+
+- Avoid creating new files on log roll during startup
+- Avoid replaying segments during startup
+- No memory overhead from storing segment indices, list of segments, etc.
+- No raft heartbeats for hibernated topics
+- No data for hibernated topics in node health telemetry messages
+- No prometheus output for topic-tagged metrics
+
+And the resources we still consume for hibernated topics:
+- Disk space for any segments that had not expired before we hibernated
+- Memory for the topic's frozen metadata
+- Bandwidth to awake and move the data on rebalance or node decommission
+
+## Interaction with other features
+
+Any 'scrub' feature will probably end up waking partitions up to
+scrub them, but that is manageable as long as the scrub mechanism
+properly limits the number of partitions to scrub at once, and groups
+the partitions together by partition when scrubbing them.
+
+Partition movement (especially during node decommission) will require
+waking up the topic.  This should not create a thundering herd during
+node evacuation, because partition movement is already scheduled to 
+avoid moving all at once.
+
+Any continuous data rebalancing feature that we implement should
+prefer to avoid migrating hibernated topics, although it can wake
+them up if they're the only viable candidates to resolve resource
+imbalances.
+
+All internal partitions will be exempt from hibernation.
+
+Partition health monitor data will include an extra field for
+last access time of a partition, or perhaps just the most-recent
+access per topic on each node, so that the controller can use this
+data to figure out the global per-topic access time to enforce
+hibernation timeout.
+
+## Telemetry & Observability
+
+- Gauge for number of topics hibernated
+- Gauge for number of partitions hibernated
+- Counter for number of hibernate events
+- Counter for number of wake events
+- rpk commands for listing topics by status
+
+## Detailed design
+
+### List of controller commands
+### List of RPCs
+
+## Corner cases dissected by example.
+
+## Drawbacks
+
+### Resource determinism & what happens if all partitions go active?
+
+Redpanda's resource footprint
+will grow and shrink depending on the number of partitions that are
+idle.  We can mitigate that somewhat if we set a ceiling on the number
+of partitions that may be active, and force-sleep the least recently used
+partitions when we hit the limit, but that doesn't help if all the partitions
+in the system really are active (we'd just be flapping them in and out of
+hibernation).
```

Comment:
An alternative model is once the limit is reached, forbid waking any new partitions up and instead return an error code. Any system in front can use this error as a signal to either scale up the cluster or migrate some topic.